src/time_bucket.c

```diff
@@ -0,0 +1,56 @@
+#include <postgres.h>
+#include <fmgr.h>
+
+#include "compat.h"
+
+#define TIME_BUCKET(period, timestamp, min, result)			\
+	do															\
+	{															\
+    if (period <= 0) \
+		ereport(ERROR, \
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE), \
+				 errmsg("period must be greater then 0"))); \
+		*(result) = (timestamp / period) * period;				\
+		if (timestamp < 0)										\
+			if (timestamp % period)								\
+			{													\
+				if (*(result) < min+ period)				\
+					*(result) = min;						\
```

Comment:
should we error here, as per discussion?

Reply:
done